jni/src/faiss_wrapper.cpp

```diff
@@ -85,6 +85,220 @@ bool isIndexIVFPQL2(faiss::Index * index);
+    // TODO: parse the entire map and remove jni object
+    std::unordered_map<std::string, jobject> subParametersCpp;
+    if(parametersCpp.find(knn_jni::PARAMETERS) != parametersCpp.end()) {
+        subParametersCpp = jniUtil->ConvertJavaMapToCppMap(env, parametersCpp[knn_jni::PARAMETERS]);
+    }
+    // end parameters to pass
+
+    // Create index
+    return indexService->initIndex(jniUtil, env, metric, indexDescriptionCpp, dim, numDocs, threadCount, subParametersCpp);
+}
+/*
+jlong knn_jni::faiss_wrapper::InitIndexFromTemplate(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jlong numDocs, jint dimJ,
+                                         jobject parametersJ, jbyteArray templateIndexJ, IndexService* indexService) {
+
+    if(dimJ <= 0) {
+        throw std::runtime_error("Vectors dimensions cannot be less than or equal to 0");
+    }
+
+    if (parametersJ == nullptr) {
+        throw std::runtime_error("Parameters cannot be null");
+    }
+
+    // parametersJ is a Java Map<String, Object>. ConvertJavaMapToCppMap converts it to a c++ map<string, jobject>
+    // so that it is easier to access.
+    auto parametersCpp = jniUtil->ConvertJavaMapToCppMap(env, parametersJ);
+
+    // Parameters to pass
+    // Metric type
+    jobject spaceTypeJ = knn_jni::GetJObjectFromMapOrThrow(parametersCpp, knn_jni::SPACE_TYPE);
+    std::string spaceTypeCpp(jniUtil->ConvertJavaObjectToCppString(env, spaceTypeJ));
+    faiss::MetricType metric = TranslateSpaceToMetric(spaceTypeCpp);
+    jniUtil->DeleteLocalRef(env, spaceTypeJ);
+
+    // Dimension
+    int dim = (int)dimJ;
+
+    // Number of docs
+    int docs = (int)numDocs;
+
+    // Index description
+    jobject indexDescriptionJ = knn_jni::GetJObjectFromMapOrThrow(parametersCpp, knn_jni::INDEX_DESCRIPTION);
+    std::string indexDescriptionCpp(jniUtil->ConvertJavaObjectToCppString(env, indexDescriptionJ));
+    jniUtil->DeleteLocalRef(env, indexDescriptionJ);
+
+    // Thread count
+    int threadCount = 0;
+    if(parametersCpp.find(knn_jni::INDEX_THREAD_QUANTITY) != parametersCpp.end()) {
+        threadCount = jniUtil->ConvertJavaObjectToCppInteger(env, parametersCpp[knn_jni::INDEX_THREAD_QUANTITY]);
+    }
+
+    // Extra parameters
+    // TODO: parse the entire map and remove jni object
+    std::unordered_map<std::string, jobject> subParametersCpp;
+    if(parametersCpp.find(knn_jni::PARAMETERS) != parametersCpp.end()) {
+        subParametersCpp = jniUtil->ConvertJavaMapToCppMap(env, parametersCpp[knn_jni::PARAMETERS]);
+    }
+    // end parameters to pass
+
+    // Create index
+    return indexService->initIndex(jniUtil, env, metric, indexDescriptionCpp, dim, numDocs, threadCount, subParametersCpp);
+}
+*/
+void knn_jni::faiss_wrapper::InsertToIndex(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jintArray idsJ, jlong vectorsAddressJ, jint dimJ,
+                                         jlong index_ptr, jobject parametersJ, IndexService* indexService) {
```

Comment:
Would it not be used at least for the thread count? I could just pass thread count though.

Reply:
You have 2 options here:
1. Pass the whole parametersJ and then only parse the threadCount out of it.
2. Pass only threadCount. 

I think 1 is better here since we need threadCount.

Reply:
Okay. I was thinking more of doing option 2 and parsing threadCount in the java part of the JNI layer, though. This was one of the optimizations I was planning on doing later since every time we insert to the index, we have to recreate the entire Java map in c++.